### ISSUE TITLE
tls: oversized record received while trying docker pull

### ISSUE THREAD

**Issue by bcgshyam** [moby/moby#37102]
**Description**
I am running a docker binary I built using [this](http://oyvindsk.com/writing/docker-build-from-source), and when I try to pull images, I get this error. 

> Error response from daemon: Get https://registry-1.docker.io/v2/: proxyconnect tcp: tls: oversized record received with length 20527

I have done the necessary proxy settings, and running docker using the default binary has no such issues. I can pull and run images without any hassle. But I get the issue while running the new binary. What changes do I need to do in order to get the pull to work behind proxy? 

**Steps to reproduce the issue:**
1. git clone https://github.com/docker/docker.git
2. cd docker
2. make binary
3. ./bundles/< path >/dockerd
4. docker pull hello-world

**Describe the results you received:**
Error response from daemon: Get https://registry-1.docker.io/v2/: proxyconnect tcp: tls: oversized record received with length 20527

**Describe the results you expected:**
Using default tag: latest
latest: Pulling from library/hello-world
Digest: sha256:f5233545e43561214ca4891fd1157e1c3c563316ed8e237750d59bde73361e77
Status: Image is up to date for hello-world:latest

**Additional information you deem important (e.g. issue happens only occasionally):**
Happens always when I am behind office proxy. Does not happen at home. 

**Output of `docker version`:**
```
Client:
 Version:      18.03.1-ce
 API version:  1.37
 Go version:   go1.9.5
 Git commit:   9ee9f40
 Built:        Thu Apr 26 07:17:20 2018
 OS/Arch:      linux/amd64
 Experimental: false
 Orchestrator: swarm

Server:
 Engine:
  Version:      dev
  API version:  1.37 (minimum version 1.12)
  Go version:   go1.10.2
  Git commit:   f4ebcb4
  Built:        Thu May 17 13:25:43 2018
  OS/Arch:      linux/amd64
  Experimental: false

```

**Output of `docker info`:**

```
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 0
Server Version: dev
Storage Driver: vfs
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 773c489c9c1b21a6d78b5c538cd395416ec50f88
runc version: 4fc53a81fb7c994640722ac585fa9ca548971871
init version: 949e6fa (expected: fec3683)
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.13.0-41-generic
Operating System: Ubuntu 16.04.4 LTS
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 3.83GiB
Name: ubuntu
ID: VSRU:CVYT:B6HQ:CABK:BHFQ:BNRD:PA52:R4C7:L7GK:PDOA:OA3N:XQO7
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
HTTP Proxy: http://web-proxy.ind.hp.com:8080
HTTPS Proxy: https://web-proxy.ind.hp.com:8080
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8

```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Ubuntu 16.04 running on VMware Workstation

**Comment by cpuguy83** [comment]
What is `proxyconnect` coming from? It sounds like it is expecting doing raw HTTP but the response was TLS.

**Comment by bcgshyam** [comment]
As I said, I am behind corporate proxy, so it might be due to that. I am not sure from where exactly it is coming from. I have followed [this](https://docs.docker.com/config/daemon/systemd/#httphttps-proxy) to setup proxy for docker. Also, I am running a new binary, after building it, so I am not sure if these settings are being used.

**Comment by mvamax** [comment]
@bcgshyam  Did you find a fix, i have exactly the same error and i am also behind an enterprise proxy.

**Comment by bcgshyam** [comment]
@mvamax No, I have not found a solution, but it worked fine for one of my colleagues who had modified the DNS in the daemon.json file, with the company DNS.  Somehow the same thing did not work for me. You can probably try that.

**Comment by bcgshyam** [comment]
Nope, haven't really tried off late, as mentioned one of my colleagues got it to work (surprisingly he didn't face this issue) and we just continued with that. Waiting for a solution.

**Comment by kevinimbrechts** [comment]
Hi,
@thaJeztah I'm not sure because it was working with Docker version 17.09.0-ce, build afdb6d4. I just updated Docker CE with `yum update` in CentOS 7 and now I have Docker version 18.06.0-ce, build 0ffa825 and I have this error message too :
`Error response from daemon: Get https://registry-1.docker.io/v2/: proxyconnect tcp: tls: oversized record received with length 20527`

**Comment by Karotta** [comment]
In my case the mistake was that I set https protocol in the https proxy instead of http protocol. Changing it in /etc/systemd/system/docker.service.d/  to "HTTPS_PROXY=http://user:pwd@proxy:port" fixed my error

**Comment by kevinimbrechts** [comment]
Oh yes thanks Karotta. My problem was the same with this version.

**Comment by mvamax** [comment]
Same problem for me too thx!

**Comment by bcgshyam** [comment]
@Karotta yep that solution works
Thanks!

**Comment by AshwinParmar** [comment]
Could you please explain this solution for Windows 10 User. How can I configured proxy settings for docker within windows 10 Enterprise behind the enterprise proxy server.

**Comment by abdullahkhilji** [comment]
Wouldn't the general proxy settings ([https://github.com/abdullahkhilji/Unix-Cheatsheet](url)) for Linux work (if the proxy settings for the enterprise are not available via docker settings.)

**Comment by mballon** [comment]
I resolved this by removing the https_proxy directive from docker config.

non-working setup:
http_proxy=http://x.x.x.x:3128
https_proxy=https://x.x.x.x:3128

working setup:
http_proxy=http://x.x.x.x:3128

**Comment by souravpatnaik** [comment]
@mballon your suggestion worked for me, thanks.
using docker version: 18.09.2

**Comment by mbheemappa** [comment]
Got the proxy configured by running below cmd.

#systemctl show --property=Environment docker
Environment=GOTRACEBACK=crash DOCKER_HTTP_HOST_COMPAT=1 PATH=/usr/libexec/docker:/usr/bin:/usr/sbin **HTTP_PROXY=http://xxxxx:8080**/ **HTTPS_PROXY=https://xxxxx:8080/**

Fix: Modifying the file /etc/systemd/system/docker.service.d/https-proxy.conf as below fixed it:

https_proxy=**http**://x.x.x.x:8080

**Comment by yan9yu** [comment]
Hi guys, I am currently using Docker on Windows and I got the same error as well. I've tried 
1. configuring the proxy in Settings 
2. creating a https-proxy.conf file under C:\Users\<My-User>\.docker 

and none of them works to me. 


```
C:\> docker run hello-world
Unable to find image 'hello-world:latest' locally
C:\Program Files\Docker\Docker\Resources\bin\docker.exe: Error response from daemon: Get https://registry-1.docker.io/v2/: proxyconnect tcp: tls: oversized record received with length 20527.
```

The following is the `docker info` output
```
C:\> docker info
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 0
Server Version: 18.09.2
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 9754871865f7fe2f4e74d43e2fc7ccd237edcbce
runc version: 09c8266bf2fcf9519a651b04ae54c967b9ab86ec
init version: fec3683
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.9.125-linuxkit
Operating System: Docker for Windows
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 1.934GiB
Name: linuxkit-00155d40c201
ID: J72T:KS7C:P2WM:R35I:WAVQ:E5RZ:7S5X:J5UT:BLT2:4T3P:WG7F:62PD
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 22
 Goroutines: 46
 System Time: 2019-05-15T03:42:04.0405259Z
 EventsListeners: 1
HTTP Proxy: https://xxxxx:xxxxx@XX.XX.XX.XX
HTTPS Proxy: https://xxxxx:xxxxx@XX.XX.XX.XX
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
Product License: Community Engine
```

**Comment by thaJeztah** [comment]
looks like you configured a https proxy to proxy http?

```
HTTP Proxy: https://xxxxx:xxxxx@XX.XX.XX.XX

**Comment by SergeyBaikal** [comment]
Nice! Thanks, Karotta!

**Comment by Wordsgames** [comment]
Hi, can you tell me that the username and password come from the server or docker session?

**Comment by fuzzy76** [comment]
Is this expected behaviour, or should it be reclassified as a bug? Can't docker communicate with a https proxy using https?
